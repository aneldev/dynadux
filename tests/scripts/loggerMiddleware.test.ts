import "jest";
import {
  createStore,
} from "../../src";

interface ITodoAppState {
  todos: ITodo[];
}

interface ITodo {
  id: string;
  label: string;
  done: boolean;
}

const actions = {
  ADD_TODO: 'ADD_TODO',
};

describe('Dynadux', () => {
  test('Middlewares', async (done) => {
    const createTodoAppStore = (onChange?: (state: ITodoAppState) => void) => {
      const store = createStore<ITodoAppState>({
        initialState: {
          todos: [],
        },
        middlewares: [
          false,
          undefined,
          null,
          {
            after: ({action, payload}) => {
              console.log('dispatch', new Date, action, payload);
            },
          }
        ],
        onChange,
        reducers: {
          [actions.ADD_TODO]: ({state: {todos}, payload}) => {
            return {
              todos: todos.concat(payload),
            };
          },
        },
      });

      return {
        get state() {
          return store.state;
        },
        addTodo: (todo: ITodo) => store.dispatch<ITodo>(actions.ADD_TODO, todo),
      };
    };

    const todoAppStore = createTodoAppStore();

    const cl = console.log;
    console.log = jest.fn();

    todoAppStore.addTodo({id: '301', label: 'Before work beers', done: false});
    todoAppStore.addTodo({id: '302', label: 'Evening beer meeting', done: false});
    todoAppStore.addTodo({id: '303', label: 'After work beers', done: false});

    expect(
      (console.log as any)
        .mock
        .calls
        .map((i: any[]) => i.filter(i => !(i instanceof Date)))
    ).toMatchSnapshot();

    console.log = cl;

    done();
  });
});
