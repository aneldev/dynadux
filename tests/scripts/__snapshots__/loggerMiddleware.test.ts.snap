// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynadux Middlewares 1`] = `
Array [
  Array [
    "dispatch",
    "ADD_TODO",
    Object {
      "done": false,
      "id": "301",
      "label": "Before work beers",
    },
  ],
  Array [
    "dispatch",
    "ADD_TODO",
    Object {
      "done": false,
      "id": "302",
      "label": "Evening beer meeting",
    },
  ],
  Array [
    "dispatch",
    "ADD_TODO",
    Object {
      "done": false,
      "id": "303",
      "label": "After work beers",
    },
  ],
]
`;
